Milestone 5

damiecode opened the pull request:
Thank you for taking the time to review our project.
Work-done
- [x] Create friendships controller
- [x] Create friendships model
- [x] Add an association between friendships and users
- [x] Ensure a user can request, accept , cancel and destroy friendships
- [x] Ensure a user cannot add himself
- [x] Add integration and unit testing for friendships

adriaanbd commented on config/routes.rb:
```diff
@@ -1,7 +1,11 @@
    devise_for :users, controllers: { registrations: 'users/registrations' }
    root to: "posts#index"
 
    resources :posts, only: [:create, :destroy, :index, :show,:new]
```
Whenever you are specifying more than half of the 7 default routes provided with `resources :model`, it could be leaner / cleaner to use the `except:` option instead, which specifies a route or list of routes that Rails should not create. Read about it [here](https://guides.rubyonrails.org/routing.html#restricting-the-routes-created).

This line could have been: `resources :posts, except: [:edit, :update]`.

[`OPTIONAL`]

adriaanbd commented on config/routes.rb:
```diff
@@ -1,7 +1,11 @@
    resources :users
+   resources :likes, only: [:create]
+   resources :comments, only: [:create]
+   resources :friendships, only: [:create, :update, :destroy, :index]
```
Whenever you are specifying more than half of the 7 default routes provided with `resources :model`, it could be leaner / cleaner to use the `except:` option instead, which specifies a route or list of routes that Rails should not create. Read about it [here](https://guides.rubyonrails.org/routing.html#restricting-the-routes-created).

This line could have been: `friendships :posts, except: [:new, :edit, :show]`.

[`OPTIONAL`]

adriaanbd commented on app/views/posts/_feed.html.erb:
```diff
@@ -1,6 +1,46 @@
+    <%if current_user == u || current_user.friend?(u)%>
+        <% u.posts.each do |p| %>
+          <li class='card mb-4 bg-white' style="width: 35rem;">
+          <div class='card-body mx-auto'>
```
This line has to be indented, and so does every line underneath it.

Furthermore, this would have been a good place to use a partial for `_post`, as this would make this `_feed` more digestible, modular, maintainable by enforcing separation of concerns.

[`OPTIONAL`]

adriaanbd commented on app/views/posts/_feed.html.erb:
```diff
@@ -1,6 +1,46 @@
+              <%end%>
+            </div>
+            <div class='row'>
+              <%= form_for(:comment, :url => { :action => :create, :controller => :comments } )  do |f| %>
```
This could have also been a good place for a `_new_comment` partial [`OPTIONAL`]

adriaanbd commented on app/views/posts/_feed.html.erb:
```diff
@@ -1,6 +1,46 @@
+              <% end %>
+              <% if liked?(p.id) %>
+                <%= button_to 'Like', likes_path(like: {user_id: current_user.id,post_id: p.id}), method: :post%>
+              <%end%>
```
Good candidate for a `_like` partial [`OPTIONAL`]

adriaanbd commented on app/views/posts/_feed.html.erb:
```diff
@@ -1,6 +1,46 @@
+                    <li class='text-dark'><%= c.content%></li>
+                  <%end%>
+                </ul> 
+              <%end%>
```
Good candidate for a `_comment` partial [`OPTIONAL`]

adriaanbd commented on app/views/posts/_feed.html.erb:
```diff
@@ -1,6 +1,46 @@
-    <p><%= p.content %></p>
+  <ul>
+  <% @users.each do |u|%>
+    <%if current_user == u || current_user.friend?(u)%>
```
Please note that `User.friend?(user)` is a Model method and shouldn't really be called by your views, as this is what `app/helpers` are for. There should be **no** application logic in the views nor helpers. Everything connected with data processing should be encapsulated in models and used in controllers. This is an additional Microverse requirement [`REQUIRED`].

For more information see [here](https://codefol.io/posts/Where-Do-I-Put-My-Code/) and [here](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers).

adriaanbd commented on app/views/users/_users.html.erb:
```diff
@@ -1,3 +1,40 @@
+<ul>
+  <% @users.each do |u| %>
+    <li class='text-dark'><%= u.name %></li><br>
+    <% if current_user.pending_friends.include?(u)%>
```
Please note that `User.pending_friends` is a Model method and shouldn't really be called by your views, as this is what `app/helpers` are for. There should be **no** application logic in the views nor helpers. Everything connected with data processing should be encapsulated in models and used in controllers. This is an additional Microverse requirement [`REQUIRED`].

For more information see [here](https://codefol.io/posts/Where-Do-I-Put-My-Code/) and [here](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers).

adriaanbd commented on app/models/user.rb:
```diff
@@ -5,6 +5,39 @@ class User < ApplicationRecord
+  has_many :friendships
+  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
+
+  def friends
```
Where is this method used in your app? I haven't seen it used yet, perhaps I'm missing something or you'll use it in the next milestone [`OPTIONAL`].

adriaanbd commented on app/models/user.rb:
```diff
@@ -5,6 +5,39 @@ class User < ApplicationRecord
+    inverse_friendships.map { |friendship| friendship.user unless friendship.confirmed }.compact
+  end
+
+  def confirm_friend(user)
```
Where is this method used in your app? I haven't seen it used yet, perhaps I'm missing something or you'll use it in the next milestone. I thought you would've used it in your friends_controller but I didn't see it [`OPTIONAL`].

damiecode commented:
Thanks for the review @ebukaume, the log out works fine on me and my coding partners system, and concerning the README if you check the example which was given by your colleague, Its exactly what we implemented, Do you need us to write the whole process of the project in the README?

ebukaume commented:
> Thanks for the review @ebukaume, the log out works fine on me and my coding partners system, and concerning the README if you check the example which was given by your colleague, Its exactly what we implemented, Do you need us to write the whole process of the project in the README?

There are some essential things you should have the README. Things like; how would a new user set up and run the app? What are the users expectations? 

As regards the signout feature; are you sure you pushed the right code? Running `rails routes` shows you have a route like this:

> `destroy_user_session DELETE /users/sign_out(.:format)                                                                devise/sessions#destroy`

The route, as you can see, matches `/users/:id`. That is why your app tries to resolve a user with id of `sign_out` as seen below.

![image](https://user-images.githubusercontent.com/22497270/71376789-e4c1a300-25c2-11ea-8ee6-dc2c3235ae14.png)

damiecode commented:
Thank you @ebukaume , I have updated the README and i was thinking since the project is not completed the README would not really be comprehensive. I also have a question to ask. My coding partner and I realized we implemented the mutual friendship for friendship version 2 in version 1, can we submit the milestone-6 like that or what can we do. Thanks

samgaco commented on spec/controllers/friendships_controller_spec.rb:
```diff
@@ -40,7 +40,10 @@
+      accept_alert do
+        click_on 'Accept'
+      end
+      have_content 'Request Accepted!'
```
:+1:
